### ISSUE TITLE
Multiple sites, no way to detect correctly which site to use

### ISSUE THREAD

**Issue by salvadormrf** [torchbox/wagtail#1283]
When using multiple sites A and B. 
A is default, port 80, www.examplea.com, points to page /home/
B, port 80, www.exampleb.com, points to page /home/

When using the page.full_url property wagtail looks for any site that matches the URL path.

As example, by generating the /sitemap.xml
www.examplea.com - > the urls point to domain from A
www.exampleb.com - >  the urls point to domain from A

In my case: the reason to have A and B pointing to the same page, is that B is just the final static domain (generated using django medusa). So we need the different domains to be generated on  sitemap.xml and share metadata where full urls are required.

In this edge case, there is any suggestion ?

**Comment by gasman** [comment]
Related: #1282

This is tricky. Making `full_url` conditional on other factors (such as 'is this medusa or not') is likely to massively complicate the URL logic throughout Wagtail. On the other hand, this seems similar to the common problem of people copying their database between live and staging sites, and forgetting to update the site record. Effectively, what we need here is different site setups for different environments (where 'live' and 'staging' are different environments, and in this case 'medusa' would be one too).

Arguably, the current sites implementation breaks one of our guiding principles for Wagtail, that developer-facing configuration should be done in code, not by clicking through admin interfaces. So, as a half-baked idea (and without thinking about how we'd migrate to this model), how about this:

* The settings file contains a section like:

        WAGTAIL_SITES = {
            'main': 'http://example.com',
            'blog': 'http://blog.example.com',
        }
* The developer can maintain separate versions of this list in settings/production.py, settings/staging.py, settings/medusa.py as appropriate
* The 'sites' admin area is then responsible for mapping these site identifiers to root pages, rather than working with the hostnames directly.

**Comment by salvadormrf** [comment]
@gasman Thanks for replying so quickly :)
I will see if we can use different settings when generating the static website.

More details:
In this case we have different environments (dev, test, stage, live) on App Engine.
All environments have a dynamic CMS (site A) and a public facing static-generated website (site B).

From site A we generate (w background task) a static website of B to be deployed/updated.

**Comment by kaedroho** [comment]
Good idea @gasman! The hostname of the site is an environment-specific thing so I think it makes sense to put it in the settings file on the server. I wonder if it's possible to do that with Djangos site framework.

**Comment by timheap** [comment]
@kaedroho It kind of is. Assuming you start from the same database for the project, you can add a number of sites: `dev.example.com`, `stage.example.com`, and `www.example.com`. Then in each deployment, you select which site to use by setting the `SITE_ID`. In previous projects where I have done this, I set `SITE_ID = 1` to be live, `SITE_ID = 2` to be stage, and `SITE_ID = 3` to be development for all the projects, so I could easily choose which behaviour I wanted.

**Comment by SalahAdDin** [comment]
@timheap Can you put a small example of your work?

**Comment by bretth** [comment]
Just lurking - not a wagtail user (yet), but I believe you want to be using django-hosts. I currently use it to manage a multi-tenant multi-domain setup.  Using a hosts.py file and a HostsMiddleware it's pretty much the same as a url pattern but for domains, and sets the desired django sites site_id in the request/response cycle. You still have access to use the full domain url via django-hosts template tags.

**Comment by jorge-marques** [comment]
:+1:  for having this configured in the settings file (even if it just overrides the db)

**Comment by mrharpo** [comment]
I just came across this problem in converting our single site [Wagtail application](https://github.com/WGBH-MLA/ov-wag/pull/258) to multi-site. I managed to get things working by doing the following:

- Configure [django-hosts](https://github.com/jazzband/django-hosts) to map a regex host name to a standardized host name
  - In our case, anything that starts with `aapb` gets mapped to the `aapb` host (e.g. `aapb-pr-123`)
  - This can be set from an environment variable, for cases where it needs to be changed for a dev/staging deployment

### `hosts.py`
```py
aapb_pattern = env.get('AAPB_HOST_PATTERN', r'aapb(.+)*')
ov_pattern = env.get('OV_HOST_PATTERN', r'ov(.+)*')

host_patterns = patterns(
    '',
    host(aapb_pattern, 'ov_wag.urls', name='aapb'),
    host(ov_pattern, 'ov_wag.urls', name='ov'),
)

```
- Add a custom middleware to match the host to a site
### `middleware.py`

```py
from wagtail.models import Site


class DjangoHostsSiteMiddleware:
    """
    Middleware that sets the correct Wagtail Site based on django-hosts routing.

    This allows Wagtail to automatically filter pages and content by site without
    needing custom filtering in each API viewset or view.
    """

    def __init__(self, get_response):
        self.get_response = get_response

    def __call__(self, request):
        # Get the hostname from the request
        hostname = request.host.name

        # Try exact match first (most efficient)
        site = Site.objects.filter(hostname=hostname).first()

        if not site:
            # Fallback to the default site (if there is one)
            site = Site.objects.filter(is_default_site=True).first()

        if site:
            # Set the site on the request so Wagtail uses it
            request.site = site
            # Also set _wagtail_site for compatibility
            request._wagtail_site = site

        response = self.get_response(request)
        return response

```

### `settings/base.py`
Enable the middleware after `django_hosts`

```py
...
MIDDLEWARE = [
    'django_hosts.middleware.HostsRequestMiddleware',
    'ov_wag.middleware.DjangoHostsSiteMiddleware',
    ... # Rest of middlewares
    'django_hosts.middleware.HostsResponseMiddleware',
]
...
```
